zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/X509ZNodeGroupAclProvider.java

```diff
@@ -0,0 +1,180 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server.auth.znode.groupacl;
+
+import java.security.cert.X509Certificate;
+import java.util.HashSet;
+import java.util.Set;
+import javax.net.ssl.X509KeyManager;
+import javax.net.ssl.X509TrustManager;
+import javax.security.auth.x500.X500Principal;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.common.ZKConfig;
+import org.apache.zookeeper.data.Id;
+import org.apache.zookeeper.server.ServerCnxn;
+import org.apache.zookeeper.server.ZooKeeperServer;
+import org.apache.zookeeper.server.auth.ServerAuthenticationProvider;
+import org.apache.zookeeper.server.auth.X509AuthenticationUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A ServerAuthenticationProvider implementation that does both authentication and authorization for protecting znodes from unauthorized access.
+ * Znodes are grouped into domains according to their ownership, and clients are granted access permission to domains.
+ * Authentication mechanism is same as in X509AuthenticationProvider. If authentication is failed,
+ *  decline the connection request.
+ * Authorization is done by checking clientId which is usually an URI (uniform resource identifier)in ACL metadata for matched domains.
+ * Detailed step for authorization is:
+ *  1. Acl provider attempts to extract the clientId from the cert provided by the client
+ *  2. Acl provider attempts to look up the client's domain using the clientId and ClientUriDomainMappingHelper
+ *  3. If matched domain is found, add the domain in authInfo in the connection object;
+ *     if no matched domain is found, add the clientId in authInfo instead; either way, establish the connection.
+ * This class is meant to support the following use patterns:
+ *  1. Single domain: The creator and the accessor to the znodes belong to the same domain. This is
+ *     the most straightforward use case, and accessors from other domains will be declined to access the nodes.
+ *  2. Super user domain: The accessors need permission to access znodes created by creators belong
+ *     to many other domains. In this case the accessors will be mapped to super user domain and be
+ *     given super user privilege.
+ *  3. Open read access: The znodes need to be accessed by accessors from many different domains, so when
```

Comment:
Careful. you're using "access" to mean crwd above, and now what you mean is only read access.